### PR TITLE
Added RectangularCuboid Primitive

### PR DESCRIPTION
**Pull request by Njord201**
I've added the Rectangular Cuboid Primitive.

For build a Rectangular Cuboid primitive, go in a .cfg file in primitives part, you can create a `rectangular_cuboid` container. All the rectangular cuboid follow this build : minX, minY, minZ, maxX, maxY and maxZ. This is the position of extrem point of the Rectangular Cuboid primitive.

You have a result like this: 
image
 For this configuration:
image

### CODE REVIEW

**Review by theophile-jr — DISMISSED** [review]

File: src/Core/Parser/Scene.cpp [hunk 1]
```diff
@@ -310,6 +310,47 @@ int Raytracer::Scene::_parsePrimitiveSetting(const libconfig::Setting &primitive
...
+            libconfig::Setting& material = rectangularCuboidArray[index].lookup("material");
+            material.lookupValue("type", materialType);
+
+            if (materialType == "flatColor") {
+                libconfig::Setting& color = material.lookup("color");
+                std::shared_ptr<FlatColor> materialPtr = std::make_shared<FlatColor>(color["r"], color["g"], color["b"]);
+                newRectangularCuboid->setMaterial(materialPtr);
+            }
+
+            // if (rectangularCuboidArray[index].exists("translation")) {
```
Comment (mathieurobert1) [inline src/Core/Parser/Scene.cpp:343]: Handle the rotation and the translation

File: src/Math/Vector3D.cpp [hunk 2]
```diff
@@ -132,3 +132,23 @@ Math::Vector3D& Math::Vector3D::operator/=(double n)
...
+
+bool Math::Vector3D::operator<(const Vector3D& ptr)
+{
+    return _vector[0] < ptr._vector[0] && _vector[1] < ptr._vector[1] && _vector[2] < ptr._vector[2];
+}
+
+bool Math::Vector3D::operator>(const Vector3D& ptr)
+{
+    return _vector[0] > ptr._vector[0] && _vector[1] > ptr._vector[1] && _vector[2] > ptr._vector[2];
+}
```
Comment (mathieurobert1) [inline src/Math/Vector3D.cpp:154]:
```suggestion
}

```

File: src/Plugins/Primitives/RectangularCuboid/RectangularCuboid.cpp [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    if (hitPoint.y() == _maxY)
+        return Math::Vector3D(0, 1, 0);
+    if (hitPoint.y() == _minY)
+        return Math::Vector3D(0, -1, 0);
+    if (hitPoint.z() == _maxZ)
+        return Math::Vector3D(0, 0, 1);
+    if (hitPoint.z() == _minZ)
+        return Math::Vector3D(0, 0, -1);
+    return Math::Vector3D(-1, -1, -1);
+}
```
Comment (mathieurobert1) [inline src/Plugins/Primitives/RectangularCuboid/RectangularCuboid.cpp:163]:
```suggestion
}

```

File: src/Plugins/Primitives/RectangularCuboid/getInstance.cpp [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+        std::cout << "New Rectangular Cuboid created ..." << std::endl;
+        return rectangularCuboid;
+    }
+
+    __attribute__((destructor))
+    static void destroysharedlibrary()
+    {
+        std::cout << "Closing Rectangular Cuboid Plugin ..." << std::endl;
+    }
+}
```
Comment (mathieurobert1) [inline src/Plugins/Primitives/RectangularCuboid/getInstance.cpp:32]:
```suggestion
}

```

**Review by mathieurobert1 — CHANGES_REQUESTED** [review]
Handle the rotation and the translation

File: include/Parser/Factory.hpp [hunk 5]
```diff
@@ -20,6 +20,7 @@
 #include "Primitives/Cone.hpp"
 #include "Primitives/Cylinder.hpp"
 #include "Primitives/Plane.hpp"
+#include "Primitives/RectangularCuboid.hpp"
 #include "Lights/Point.hpp"
 #include "Lights/Ambient.hpp"
 #include "Lights/Directional.hpp"
```
Comment (AxelF44) [inline include/Parser/Factory.hpp:26]:
```suggestion
#include "Lights/Point.hpp"
#include "Lights/Ambient.hpp"
#include "Primitives/Cone.hpp"
#include "Primitives/Plane.hpp"
#include "Lights/Directional.hpp"
#include "Primitives/Cylinder.hpp"
#include "Primitives/RectangularCuboid.hpp"
```

File: include/Primitives/RectangularCuboid.hpp [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+            /**
+             * @brief Construct a new RectangularCuboid object.
+             *
+             * @param maxX double max x of the rectangular cuboid
+             * @param maxY double max y of the rectangular cuboid
+             * @param maxZ double max z of the rectangular cuboid
+             * @param minX double min x of the rectangular cuboid
+             * @param minY double min y of the rectangular cuboid
+             * @param minZ double min z of the rectangular cuboid
+             */
```
Comment (AxelF44) [inline include/Primitives/RectangularCuboid.hpp:31]: Did you forget to explain this param: std::shared_ptr<Material::IMaterial> material ?

File: include/Primitives/RectangularCuboid.hpp [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+            RectangularCuboid(double maxX, double maxY, double maxZ, double minX, double minY, double minZ, std::shared_ptr<Material::IMaterial> material);
+
+            /**
+             * @brief Destroy the RectangularCuboid object
```
Comment (AxelF44) [inline include/Primitives/RectangularCuboid.hpp]:
```suggestion
             * @brief Destroy the RectangularCuboid object.
```

File: include/Primitives/RectangularCuboid.hpp [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+             * @param vDir - vector direction of the ray
+             * @param origin - origin of the ray
+             * @return hitPoint if the ray reach the primitive or
+             * @return (-1, -1, -1) if the ray doesn't reach the primitive
+             */
+            Math::Point3D checkRayReachPrimitive(double t, Math::Vector3D vDir, Math::Point3D origin) const;
+
+            double _maxX;
+            double _minX;
+            double _maxY;
```
Comment (AxelF44) [inline include/Primitives/RectangularCuboid.hpp:176]: I you align _material and rotation keep going to the rest of param please

**Review by mathieurobert1 — DISMISSED** [review]

**Review by AxelF44 — CHANGES_REQUESTED** [review]

**Review by mathieurobert1 — APPROVED** [review]

**Review by AxelF44 — APPROVED** [review]

**Pull request merged**